Cleanup EditorNode, EditorHistory, EditorSelection and EditorData

Geometror opened the pull request:
This PR aims to tidy-up the EditorNode, EditorData, EditorSelection and EditorHistory classes (partially) with the following changes:
- Remove unused enums (unused HISTORY_MAX constant)
- Remove unused class attributes
- Remove some old commented out code (reenabling it caused causes undesired behavior, so it seems fine to remove it)
- Some renames of variables to be more descriptive/explicit (e.g., the term "history" was used ambiguously)
	- Rename EditorHistory to EditorSelectionHistory to better reflect what it does (this might require some discussion, as I am not quite sure with it)
- Adjusted the comment style for consistency (// Begin with whitespace + uppercase letter and place a dot at the end.)
- Removed some uninformative comments (e.g., `//bye`, `//save` after save();)
- Restructure the EditorNode header to improve readability 
	- Clearly separate class attributes and methods
	- Move enum and struct declarations
	- Group static methods
	- Group attributes/methods in a way that makes sense (I hope)
- Reimplement rendering_driver check in EditorNode::_rendering_driver_selected (old commented out version)
- Add TODO/FIXME in front of  some comments which were (implicitly or explicitly) suggesting changes/marking hacks.
- Remove an unnecessary visibility modifier since the GDCLASS macro includes "private:" at the end
- Includes almost all changes in https://github.com/godotengine/godot/pull/44222 and https://github.com/godotengine/godot/pull/44249 except the simplifications using the ERR_FAIL_INDEX macro, therefore co-authored by @EricEzaM.

EricEzaM commented:
You may be interested in having at a look at these PRs and seeing if there are any changes worth bringing into this one (I just never got around to updating them) #44222 #44249

Geometror commented:
Rebased and incorporated some of the changes made in https://github.com/godotengine/godot/pull/44222 and https://github.com/godotengine/godot/pull/44249 (basically all of it except the ERR_FAIL_INDEX macro additions/simplifications since they were used where negative values indicate a valid state, resulting in error spamming).

akien-mga commented on editor/editor_node.cpp:
```diff
@@ -2967,7 +2967,6 @@ void EditorNode::_menu_option_confirm(int p_option, bool p_confirmed) {
 			OS::get_singleton()->shell_open("https://godotengine.org/donate");
 		} break;
-
 			//		case SET_VIDEO_DRIVER_SAVE_AND_RESTART: {
 			//			ProjectSettings::get_singleton()->set("rendering/driver/driver_name", video_driver_request);
 			//			save_all_scenes();
```
This is dead code that can be removed.

Geometror replied:
Done.

akien-mga reviewed (approved):
Looks pretty good to me. It makes variables and comments clearer, and the API renames and refactoring seems useful/clearer.

KoBeWi commented on editor/editor_node.cpp:
```diff
@@ -1393,7 +1394,8 @@ void EditorNode::_set_scene_metadata(const String &p_file, int p_idx) {
 
-	scene->set_meta("__editor_run_settings__", Variant()); // clear it (no point in keeping it)
+	// Clear meta value (no point in keeping it).
+	scene->set_meta("__editor_run_settings__", Variant());
 	scene->set_meta("__editor_plugin_states__", Variant());
```
This could be changed to `remove_meta()`

Geometror replied:
Just did a quick search and it looks like this is actually the only place where these meta values are used, so I guess we could remove them completely.

akien-mga replied:
Indeed:
```
$ rg __editor_run_settings__
editor/editor_node.cpp
1396:   scene->set_meta("__editor_run_settings__", Variant()); // clear it (no point in keeping it)
$ rg __editor_plugin_states__
editor/editor_node.cpp
1397:   scene->set_meta("__editor_plugin_states__", Variant());
```

Geometror replied:
Removed it.

KoBeWi commented on editor/editor_node.cpp:
```diff
@@ -2289,7 +2289,7 @@ void EditorNode::_edit_current(bool p_skip_foreign) {
 	}
 
-	/* Take care of PLUGIN EDITOR */
+	// Take care of PLUGIN EDITOR.
```
This comment is odd. What is PLUGIN EDITOR?

Geometror replied:
You're right, this is odd. We could change it to `// Take care of the main editor plugin.`

KoBeWi commented on editor/editor_node.cpp:
```diff
@@ -3366,14 +3365,14 @@ void EditorNode::set_addon_plugin_enabled(const String &p_addon, bool p_enabled,
 	EditorPlugin *ep = memnew(EditorPlugin);
 	ep->set_script(script);
-	plugin_addons[p_addon] = ep;
+	name_to_editor_plugin[p_addon] = ep;
```
`name_to_editor_plugin` doesn't sound right. The original name is ok, not sure how it can be improved.

akien-mga replied:
The "addon_plugin" name in the method itself is a bit weird. Doesn't `addon` and `plugin` mean the same in this context? Or is there an API for non-plugin addons?

Since this stores editor plugins, it could just be `editor_plugins`.

Geometror replied:
The point of this rename was that "addon" is inconsistent which the usual "plugin" naming, but just changing "addons" to "plugins" wont work since there is already `Vector<EditorPlugin *> editor_plugins;` 
In addition, "plugin_addons" sound like it is just a list of all editor plugins/addons, which is not the case: It is actually of the type Map<String, EditorPlugin *>, so it maps the name of the editor plugin to the editor plugin object. In my opinion naming a map x_to_y makes it clearer since it directly tells you what is mapped to what.

KoBeWi replied:
If I understand this variable correctly, it tracks which plugin was added from which user addon.

Geometror replied:
Ok, this naming is indeed a bit weird. "Addon plugins" are basically editor plugins provided by the user, right? Should we then rename the corresponding methods/variables to use a "user_plugin" name scheme instead of a mix of plugin_addons and addon_plugins?

KoBeWi replied:
I don't know tbh. It's rather established that installable plugins are called addons (their default folder is even called that). "addon plugins" could indeed be renamed to "user plugins", but the other usage of "addon" is probably fine.

KoBeWi commented on editor/editor_node.cpp:
```diff
@@ -3933,7 +3932,7 @@ void EditorNode::register_editor_types() {
 	// FIXME: Is this stuff obsolete, or should it be ported to new APIs?
 	GDREGISTER_CLASS(EditorScenePostImport);
-	//ClassDB::register_type<EditorImportExport>();
+	// ClassDB::register_type<EditorImportExport>();
```
Seems like this type no longer exists.

akien-mga replied:
If we remove the dead code I guess we can also remove the `FIXME` comment above. I added this more than a year ago, if this didn't change since it's likely not relevant anymore.

KoBeWi replied:
Isn't it for EditorScenePostImport? Or was EditorScenePostImport just added between the comment and the commented out code?

Geometror replied:
Ok, removed both comments.

Geometror replied:
In PR #10656
```
//ClassDB::register_type<EditorImporter>();
//ClassDB::register_type<EditorPostImport>();
```
was replaced with:
```
// FIXME: Is this stuff obsolete, or should it be ported to new APIs?
//ClassDB::register_class<EditorExportPlugin>();
//ClassDB::register_class<EditorScenePostImport>();
//ClassDB::register_type<EditorImportExport>();
```

KoBeWi replied:
IIRC EditorScenePostImport was ported recently. The only thing left would be EditorImportExport, but it's not clear what it's supposed to do. Also it's 5 years old...

Geometror replied:
I did a quick research with git log, but couldn't find the exact commit where EditorImportExport was removed or renamed. But in the end it should be fine to remove this commented-out line.

fire replied:
Good to remove.

Geometror commented on editor/editor_node.cpp:
```diff
@@ -6559,10 +6562,10 @@ EditorNode::EditorNode() {
 
 	if (OS::get_singleton()->get_data_path() == OS::get_singleton()->get_config_path()) {
-		// Configuration and data folders are located in the same place (Windows/macos)
+		// Configuration and data folders are located in the same place (Windows/macos).
```
Changed to "MacOS".

KoBeWi commented on editor/editor_node.h:
```diff
@@ -218,13 +221,58 @@ class EditorNode : public Node {
+
+	static EditorNode *singleton;
+
+	EditorCommandPalette *command_palette;
+	EditorData editor_data;
+	EditorExport *editor_export;
+	EditorFolding editor_folding;
+	EditorInterface *editor_interface;
+	EditorLog *log;
+	EditorNativeShaderSourceVisualizer *native_shader_source_visualizer;
+	EditorPlugin *editor_plugin_screen;
+	EditorPluginList *editor_plugins_force_input_forwarding;
+	EditorPluginList *editor_plugins_force_over;
+	EditorPluginList *editor_plugins_over;
+	EditorQuickOpen *quick_open;
+	EditorQuickOpen *quick_run;
+	EditorResourcePreview *resource_preview;
+	EditorRun editor_run;
+	EditorRunNative *run_native;
+	EditorSelection *editor_selection;
+	EditorSelectionHistory editor_history;
+	EditorSettingsDialog *editor_settings_dialog;
+
+	ProjectExportDialog *project_export;
+	ProjectSettingsEditor *project_settings_editor;
```
These could be initialized to `nullptr` here.

Geometror replied:
You're right, I changed it accordingly. But aren't there much more pointer member variables which are not initialized to nullptr (across the whole codebase)? So is there a general rule where to initialize to `nullptr` or should it ideally always be done?

KoBeWi replied:
Ideally, it should always be done. If it's not done somewhere, it's usually old code or someone missed it during review. We have occasional crashes caused by uninitialized pointers or variables.

akien-mga replied:
We did a pass at initialization most member variables a while ago, but at the time we did not decide to initialize pointers to `nullptr`. So most of the codebase doesn't initialize them, but we can confirm with the other contributors that this should be done and then implement it with a regex.

KoBeWi replied:
If you do `if (pointer)`, the behavior is undefined when the pointer is not initialized and results in crashes that occur randomly (I've seen two of them recently). I don't know if there is anything against initializing them, but I'd say it's better to be safe.

Geometror committed:
Cleanup EditorNode and EditorData

Co-authored-by: Eric M <itsjusteza@gmail.com>

akien-mga commented:
Thanks!
